server/src/daemon/secret_exchange.rs

```diff
@@ -20,55 +22,58 @@ pub struct SecretExchange {
+    // Creates the initial payload containing caller public_key
+    pub fn secret_exchange_begin(&self) -> String {
+        let map = HashMap::from([(PUBLIC, BASE64_STANDARD.encode(&self.public_key))]);
 
-        exchange
+        encode(&map)
     }
 
-    pub fn secret_exchange_receive(
-        &self,
-        exchange_in: HashMap<&str, String>,
-    ) -> Result<HashMap<&str, String>, base64::DecodeError> {
+    // Decrypt and retrieve secret
+    pub fn secret_exchange_receive(&self, exchange: String) -> Result<String, base64::DecodeError> {
+        let map = decode(&exchange);
         let mut exchange_out: HashMap<&str, String> = HashMap::new();
-        let secret = exchange_in.get(SECRET);
+
+        let secret = map.get(SECRET);
         if !secret.is_none() {
             let secret = BASE64_STANDARD.decode(secret.unwrap())?;
-            let public_key = BASE64_STANDARD.decode(exchange_in.get(PUBLIC).unwrap())?;
+            let public_key = BASE64_STANDARD.decode(map.get(PUBLIC).unwrap())?;
             let public_key = Key::new(public_key);
-            let iv = BASE64_STANDARD.decode(exchange_in.get(IV).unwrap())?;
+            let iv = BASE64_STANDARD.decode(map.get(IV).unwrap())?;
 
             let aes_key = Key::generate_aes_key(&self.private_key, &public_key);
             let secret = crypto::decrypt(secret, &aes_key, &iv).to_vec();
 
             let secret = std::str::from_utf8(&secret).unwrap().to_string();
 
-            exchange_out.insert(SECRET, secret);
+            exchange_out.insert(SECRET, secret); // decrypted password is not re
+                                                 // encoded to base64
         }
         exchange_out.insert(PUBLIC, BASE64_STANDARD.encode(&self.public_key));
 
-        Ok(exchange_out)
+        Ok(encode(&exchange_out))
     }
 
+    // Send Secret and perform encryption
     pub fn secret_exchange_send(
         &self,
         secret: &str,
-        exchange: HashMap<&str, String>,
-    ) -> Result<HashMap<&str, String>, base64::DecodeError> {
-        let public_key = BASE64_STANDARD.decode(exchange.get(PUBLIC).unwrap())?;
-        let public_key = Key::new(public_key);
+        exchange: String,
+    ) -> Result<String, base64::DecodeError> {
+        let map = decode(&exchange);
 
+        let public_key = Key::new(BASE64_STANDARD.decode(map.get(PUBLIC).unwrap())?);
         let aes_key = Key::generate_aes_key(&self.private_key, &public_key);
         let iv = crypto::generate_iv();
         let secret = crypto::encrypt(secret, &aes_key, &iv);
 
-        let mut exchange: HashMap<&str, String> = HashMap::new();
-        exchange.insert(SECRET, BASE64_STANDARD.encode(secret));
-        exchange.insert(PUBLIC, BASE64_STANDARD.encode(&self.public_key));
-        exchange.insert(IV, BASE64_STANDARD.encode(iv));
+        let map = HashMap::from([
+            (PUBLIC, BASE64_STANDARD.encode(&self.public_key)),
```

Comment:
Given we only use this KeyFile format with binary blob as values, it might be simpler to move the base64 encoding/decoding logic to the `encode`/`decode` functions, like `fn encode(map: &HashMap<&str, &[u8]>) -> Result<String, base64::DecodeError>`?

Reply:
make sense, thanks! Moved base64 encoding/decoding out of SecretExchange API and into the encode/decode functions.